glue_runner.py

```diff
@@ -0,0 +1,276 @@
+            line = line.strip()
+            if line.find("DELTACAT_PRESIGNED_URL=") != -1:
+                dc_url = line.split("=", 1)[1]
+        returncode = proc.poll()
+        if returncode != 0:
+            raise RuntimeError(f"Command {cmd} Failed. Exit Code: {returncode}")
+
+    s3_client = boto3.client("s3")
+
+    # upload the example runner script to S3
+    bucket_name = f"deltacat-glue-scripts-{stage}"
+    print(f"Ensuring Glue Job Script Bucket exists: {bucket_name}")
+    s3_client.create_bucket(Bucket=bucket_name)
+    print(f"Using Glue Job Script Bucket: {bucket_name}")
+
+    # create or update the glue job
+    glue_job_name = args.glue_job_name
+    if not glue_job_name:
+        glue_job_name = f"deltacat-example-runner-{stage}"
+    run_script_path = os.path.relpath(args.script_path)
+    s3_glue_script_file_path = f"s3://{bucket_name}/{run_script_path}"
+    print(f"Uploading script '{run_script_path}' to '{s3_glue_script_file_path}'")
+    s3_client.upload_file(
+        run_script_path,
+        bucket_name,
+        run_script_path,
+    )
+    print(f"Uploaded Glue Job Script to: {s3_glue_script_file_path}")
+    glue_max_workers = args.glue_max_workers
+    glue_iam_role = args.glue_iam_role
+    glue_job_config = {
+        "Role": glue_iam_role,
+        "GlueVersion": "4.0",
+        "ExecutionProperty": {"MaxConcurrentRuns": 10},
+        "WorkerType": "Z.2X",
+        "NumberOfWorkers": glue_max_workers,
+        "Command": {
+            "Name": "glueray",
+            "Runtime": "Ray2.4",
+            "PythonVersion": "3.9",
+            "ScriptLocation": s3_glue_script_file_path,
+        },
+    }
+    region = args.region
+    glue_client = boto3.client(
+        "glue",
+        region,
+    )
+    try:
+        response = glue_client.get_job(JobName=glue_job_name)
+        print(f"Found Existing Glue Job `{glue_job_name}`: {response}")
+        print(f"Updating Existing Glue Job `{glue_job_name}`")
+        if not glue_iam_role:
+            glue_job_config["Role"] = response["Job"]["Role"]
+        glue_client.update_job(
+            JobName=glue_job_name,
+            JobUpdate=glue_job_config,
+        )
+    except ClientError as e:
+        print(f"Glue Job `{glue_job_name}` doesn't exist. Creating it.")
+        # create or update a glue job
+        glue_client.create_job(
+            Name=glue_job_name,
+            **glue_job_config,
+        )
+    print(f"New Glue Job Config: {glue_job_config}")
+
+    print(f"Getting Last Job Run Arguments for {glue_job_name}")
+    response = glue_client.get_job_runs(
+        JobName=glue_job_name,
+    )
+    job_runs = response["JobRuns"]
+    last_job_run_args = (
+        job_runs[0]["Arguments"] if job_runs and "Arguments" in job_runs[0] else {}
+    )
+    print(f"Last Job Run Args for {glue_job_name}: {last_job_run_args}\n\n")
+
+    cli_args = args.example_script_args
+    last_job_run_packages = last_job_run_args.get("--PACKAGE")
```

Comment:
How is this different from pip-install? Wondering if we are double installing? 1/pip-install 2/runtime_env using `PACKAGE` (based on fixture.py example).

Reply:
This was an old workaround to ensure that the runtime environment was installed on all cluster nodes (including the head node) that was needed in the earlier days of the AWS Glue for Ray Preview release, but is no longer needed today. Removed (doesn't seem to impact long 1-3 minute runtime environment setup times much though, unfortunately).